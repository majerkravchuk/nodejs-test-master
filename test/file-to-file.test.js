'use strict';

const { readFileSync } = require('fs');
const { pipeline } = require('stream');
const { resolve } = require('path');

const Source = require('../transport/file-source');
const Target = require('../transport/file-target');
const LogTransform = require('../transport/log-transform');

const promisefyPipeline = (source, parser, target) => {
  return new Promise((resolve, reject) => {
    pipeline(
      source.stream,
      parser,
      target.stream,
      err => {
        if (err) {
          reject(err);
        } else {
          // I know, it's a bit ugly, but for a quick test in a test task - does its job xD
          resolve(readFileSync(target.config.path).toString());
        }
      }
    )
  });
}

describe('file-to-file', () => {
  describe('reads data from source.txt and writes parsed logs into output file', () => {
    it('transforms logs to the desired format', async () => {
      const promise = promisefyPipeline(
        new Source({
          type: 'file',
          path: resolve(__dirname, './data/source.txt')
        }),
        new LogTransform(),
        new Target({
          type: 'file',
          path: resolve(__dirname, './data/file-to-file.test-output.txt')
        })
      );
      await expect(promise).resolves.toMatchSnapshot();
    });

    it('filters logs', async () => {
      const promise = promisefyPipeline(
        new Source({
          type: 'file',
          path: resolve(__dirname, './data/source.txt')
        }),
        new LogTransform({
          types: [
            'ERROR',
            'WARNING'
          ]
        }),
        new Target({
          type: 'file',
          path: resolve(__dirname, './data/file-to-file-with-filters.test-output.txt')
        })
      );
      await expect(promise).resolves.toMatchSnapshot();
    });
  });
});

