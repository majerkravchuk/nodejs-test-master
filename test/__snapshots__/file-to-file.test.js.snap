// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file-to-file reads data from source.txt and writes parsed logs into output file filters logs 1`] = `
"{\\"logsource\\":\\"client\\",\\"env\\":\\"prod\\",\\"type\\":\\"ERROR\\",\\"timestamp\\":\\"2018-02-09T11:40:40.000Z\\",\\"message\\":\\"An application event log entry..\\",\\"_data\\":{\\"user_id\\":1}}
{\\"logsource\\":\\"server\\",\\"program\\":\\"app\\",\\"host\\":\\"127.0.0.1\\",\\"type\\":\\"ERROR\\",\\"timestamp\\":\\"2003-10-11T22:14:15.003Z\\",\\"message\\":\\"ERROR\\",\\"_data\\":{}}
{\\"logsource\\":\\"server\\",\\"program\\":\\"app\\",\\"env\\":\\"prod\\",\\"type\\":\\"WARNING\\",\\"message\\":\\"WARNING\\",\\"_data\\":{\\"info\\":{\\"some\\":\\"data\\"},\\"data\\":{\\"some\\":\\"data\\"}}}
{\\"type\\":\\"ERROR\\",\\"message\\":\\"BadRequestError: request aborted | at emitNone (events.js:105:13) | at IncomingMessage.emit (events.js:207:7) | at abortIncoming (_http_server.js:410:9) | at socketOnClose (_http_server.js:404:3) | at emitOne (events.js:120:20) | at Socket.emit (events.js:210:7) | at TCP._handle.close [as _onclose] (net.js:547:12)\\"}
{\\"type\\":\\"ERROR\\",\\"timestamp\\":\\"2018-02-09T11:40:40.000Z\\",\\"message\\":\\"BadRequestError: request aborted | at emitNone (events.js:105:13) | at IncomingMessage.emit (events.js:207:7) | at abortIncoming (_http_server.js:410:9) | at socketOnClose (_http_server.js:404:3) | at emitOne (events.js:120:20) | at Socket.emit (events.js:210:7) | at TCP._handle.close [as _onclose] (net.js:547:12)\\"}
"
`;

exports[`file-to-file reads data from source.txt and writes parsed logs into output file transforms logs to the desired format 1`] = `
"{\\"logsource\\":\\"client\\",\\"env\\":\\"prod\\",\\"type\\":\\"ERROR\\",\\"timestamp\\":\\"2018-02-09T11:40:40.000Z\\",\\"message\\":\\"An application event log entry..\\",\\"_data\\":{\\"user_id\\":1}}
{\\"logsource\\":\\"server\\",\\"program\\":\\"app\\",\\"host\\":\\"127.0.0.1\\",\\"type\\":\\"ERROR\\",\\"timestamp\\":\\"2003-10-11T22:14:15.003Z\\",\\"message\\":\\"ERROR\\",\\"_data\\":{}}
{\\"logsource\\":\\"server\\",\\"program\\":\\"app\\",\\"env\\":\\"prod\\",\\"type\\":\\"WARNING\\",\\"message\\":\\"WARNING\\",\\"_data\\":{\\"info\\":{\\"some\\":\\"data\\"},\\"data\\":{\\"some\\":\\"data\\"}}}
{\\"type\\":\\"ERROR\\",\\"message\\":\\"BadRequestError: request aborted | at emitNone (events.js:105:13) | at IncomingMessage.emit (events.js:207:7) | at abortIncoming (_http_server.js:410:9) | at socketOnClose (_http_server.js:404:3) | at emitOne (events.js:120:20) | at Socket.emit (events.js:210:7) | at TCP._handle.close [as _onclose] (net.js:547:12)\\"}
{\\"logsource\\":\\"server\\",\\"program\\":\\"app\\",\\"host\\":\\"127.0.0.1\\",\\"env\\":\\"prod\\",\\"type\\":\\"INFO\\",\\"timestamp\\":\\"2003-10-11T22:14:15.003Z\\",\\"message\\":\\"INFO\\",\\"_data\\":{\\"info\\":{}}}
{\\"logsource\\":\\"server\\",\\"program\\":\\"app\\",\\"host\\":\\"127.0.0.1\\",\\"env\\":\\"prod\\",\\"type\\":\\"INFO\\",\\"timestamp\\":\\"2003-10-11T22:14:15.003Z\\",\\"message\\":\\"DEBUG\\",\\"_data\\":{\\"info\\":{}}}
{\\"logsource\\":\\"client\\",\\"program\\":\\"client_app\\",\\"host\\":\\"127.0.0.1\\",\\"env\\":\\"prod\\",\\"type\\":\\"INFO\\",\\"timestamp\\":\\"2018-02-09T11:40:40.000Z\\",\\"message\\":\\"An application event log entry..\\",\\"_data\\":{\\"user_id\\":1}}
{\\"type\\":\\"ERROR\\",\\"timestamp\\":\\"2018-02-09T11:40:40.000Z\\",\\"message\\":\\"BadRequestError: request aborted | at emitNone (events.js:105:13) | at IncomingMessage.emit (events.js:207:7) | at abortIncoming (_http_server.js:410:9) | at socketOnClose (_http_server.js:404:3) | at emitOne (events.js:120:20) | at Socket.emit (events.js:210:7) | at TCP._handle.close [as _onclose] (net.js:547:12)\\"}
"
`;
